src/MongoDB.Driver/LibMongoCrypt/EncryptionSource.cs

```diff
@@ -0,0 +1,163 @@
+* limitations under the License.
+*/
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using MongoDB.Bson;
+using MongoDB.Bson.IO;
+using MongoDB.Crypt;
+using MongoDB.Driver.Core.Misc;
+
+namespace MongoDB.Driver.LibMongoCrypt
+{
+    /// <summary>
+    /// Represents a encryption source.
+    /// </summary>
+    public interface IEncryptionSource //: IDisposable
+    {
+#pragma warning disable 3003
+        /// <summary>
+        /// Gets the singleton instance of <see cref="CryptClient"/>.
+        /// </summary>
+        /// <returns>The singleton instance of <see cref="CryptClient"/>.</returns>
+        CryptClient CryptClient { get; }
+#pragma warning restore
+        
+        /// <summary>
+        /// Gets the mongocryptd client.
+        /// </summary>
+        IMongoClient MongoCryptDClient { get; }
+    }
+
+    internal class EncryptionSource : IEncryptionSource
+    {
+        private readonly AutoEncryptionOptions _autoEncryptionOptions;
+        private readonly Lazy<CryptClient> _cryptClient;
+        private readonly IMongoClient _mongoCryptDClient;
+
+        private EncryptionSource(MongoClientSettings mongoClientSettings)
+        {
+            Ensure.IsNotNull(mongoClientSettings, nameof(mongoClientSettings));
+            _autoEncryptionOptions = Ensure.IsNotNull(mongoClientSettings.AutoEncryptionOptions, nameof(mongoClientSettings.AutoEncryptionOptions));
+            _mongoCryptDClient = CreateMongoCryptDClient(_autoEncryptionOptions.ExtraOptions);
+            _cryptClient = new Lazy<CryptClient>(() => CreateCryptClient(CreateCryptOptions()));
+        }
+
+        public IMongoClient MongoCryptDClient => _mongoCryptDClient;
+
+#pragma warning disable 3003
+        public CryptClient CryptClient => _cryptClient.Value;
+#pragma warning restore
+
+        public static IEncryptionSource CreateEncryptionSourceIfNecessary(MongoClientSettings mongoClientSettings)
+        {
+            if (mongoClientSettings.AutoEncryptionOptions != null)
+            {
+                return new EncryptionSource(mongoClientSettings);
+            }
+            else
+            {
+                return null;
+            }
+        }
+
+        private CryptOptions CreateCryptOptions()
+        {
+            IKmsCredentials kmsCredentials = null;
+            var kmsProviders = _autoEncryptionOptions.KmsProviders;
+            if (kmsProviders != null)
+            {
+                if (kmsProviders.TryGetValue("aws", out var awsProvider))
+                {
+                    if (awsProvider.TryGetValue("accessKeyId", out var accessKeyId) &&
+                        awsProvider.TryGetValue("secretAccessKey", out var secretAccessKey))
+                    {
+                        kmsCredentials = new AwsKmsCredentials((string)secretAccessKey, (string)accessKeyId);
+                    }
+                }
+                if (kmsProviders.TryGetValue("local", out var localProvider))
```

Comment:
In case if there are two provided `kms providers`, we need to save both of them. Otherwise, some tests become failed.